base/applications/osk/main.c

```diff
@@ -116,78 +131,281 @@ DWORD WINAPI OSK_WarningDlgThread(LPVOID lpParameter)
+            {
+                szKey = Keys[i].name;
+            }
+            
+            /* Only one & the button will try to underline the next character... */
+            if (wcsncmp(szKey, L"&", 1) == 0)
+                szKey = L"&&";
+
+            SetWindowTextW(Globals.hKeys[i], szKey);
+        }
+        return 0;
+    }
+
+    /* Only font update is required */
+    if (reason == SETKEYS_FONT)
+    {
+        for (int i = 0; i < Globals.Keyboard->KeyCount; i++)
+            SendMessageW(Globals.hKeys[i], WM_SETFONT, (WPARAM)Globals.hFont, TRUE);
+
+        return 0;
+    }
+
+    /* Clear up current layout before applying a different one */
+    if (reason == SETKEYS_LAYOUT)
+        OSK_DestroyKeys();
+
+    if (Globals.bIsEnhancedKeyboard)
+    {
+        Globals.Keyboard = &EnhancedKeyboard;
+    }
+    else
+    {
+        Globals.Keyboard = &StandardKeyboard;
+    }
+
+    Globals.hKeys = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(HWND) * Globals.Keyboard->KeyCount);
+
+    if (!Globals.hKeys)
+    {
+        DPRINT("OSK_SetKeys(): Failed to allocate memory for button handles.\n");
+        return -1;
+    }
+
+    Keys = Globals.Keyboard->Keys;
+
+    /* Create key buttons */
+    for (int i = 0; i < Globals.Keyboard->KeyCount; i++)
+    {
+        uVirtKey = MapVirtualKeyW(Keys[i].scancode & SCANCODE_MASK, MAPVK_VSC_TO_VK);
+
+        if (Keys[i].translate && ToUnicode(uVirtKey, Keys[i].scancode & SCANCODE_MASK, bKeyStates, wKey, _countof(wKey), 0) >= 1)
+        {
+            szKey = wKey;
+        }
+        else
+        {
+            szKey = Keys[i].name;
+        }
+        
+        Globals.hKeys[i] = CreateWindowW(L"button",
+                                         szKey,
+                                         WS_VISIBLE | WS_CHILD | BS_PUSHBUTTON | Keys[i].flags,
+                                         Keys[i].x,
+                                         Keys[i].y,
+                                         Keys[i].cx,
+                                         Keys[i].cy,
+                                         Globals.hMainWnd,
+                                         (HMENU)Keys[i].scancode,
+                                         Globals.hInstance,
+                                         NULL);
+        if (Globals.hFont)
+            SendMessageW(Globals.hKeys[i], WM_SETFONT, (WPARAM)Globals.hFont, 0);
+    }
+
+    /* Add additional padding for caption and menu */
+    int y_pad = GetSystemMetrics(SM_CYSIZE) + GetSystemMetrics(SM_CYMENU);
+    /* Size window according to layout */
+    SetWindowPos(Globals.hMainWnd,
+                (Globals.bAlwaysOnTop ? HWND_TOPMOST : HWND_NOTOPMOST),
+                0,
+                0,
+                Globals.Keyboard->Size.cx,
+                Globals.Keyboard->Size.cy + y_pad,
+                SWP_NOMOVE);
+    
+    /* Create LEDs */
+    POINT LedPos  = Globals.Keyboard->LedStart;
+    SIZE LedSize = Globals.Keyboard->LedSize;
```

Comment:
Please respect the current style of the code where variables are declared at the top first.